scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamMode.scala

```diff
@@ -74,3 +78,54 @@ object BeamSink extends Serializable {
         pc.asInstanceOf[PCollection[String]].apply(TextIO.write().to(path))
     }
 }
+
+class BeamFileIO[T](output: String) extends BeamSink[T] {
+  override def write(
+      pipeline: Pipeline,
+      config: Config,
+      pc: PCollection[_ <: T]
+  ): Unit = {
+    val pColT: PCollection[T] = BeamFunctions.widenPCollection(pc)
+
+    pColT.apply(
+      FileIO
+        .write()
+        .via(new CoderFileSink(pColT.getCoder))
+        .to(output)
+        .withNumShards(1)
+    )
+  }
+}
+
+class CoderFileSink[T](coder: Coder[T]) extends FileIO.Sink[T] {
+  private var outputStream: java.io.OutputStream = _
+
+  override def open(channel: WritableByteChannel): Unit =
+    outputStream = Channels.newOutputStream(channel)
+
+  override def write(element: T): Unit = coder.encode(element, outputStream)
+  override def flush(): Unit = outputStream.flush()
+}
+
+class InputStreamIterator[T](stream: InputStream, coder: Coder[T]) extends Iterator[T] {
+  var hasNextRecord: Boolean = _
+  var nextRecord: T = _
+
+  fetchNext()
+  override def hasNext: Boolean = hasNextRecord
+
+  override def next(): T = {
+    val recordToReturn = nextRecord
+    fetchNext()
+    recordToReturn
+  }
+
+  private def fetchNext(): Unit =
+    Try(coder.decode(stream)) match {
```

Comment:
this worries me. What if you see an error that is not EOF?

Also, boxing with Try is an extra cost you don't need.
```scala
try {
  nextRecord = coder.decode(stream)
  hasNextRecord = true
}
catch {
  case _: EOFException =>
    hasNextRecord = false
  // don't catch other errors.
}
```